lib/mongoid/attributes.rb

```diff
@@ -86,7 +86,9 @@ def has_attribute_before_type_cast?(name)
     def read_attribute(name)
       field = fields[name.to_s]
       raw = read_raw_attribute(name)
-      field ? field.demongoize(raw) : raw
+      value = field ? field.demongoize(raw) : raw
+      attribute_will_change!(name.to_s) if value.resizable?
```

Comment:
@Neilshweky this seems like the wrong approach. The `read_attribute` method should not mutate any internal state (it is expected to be read only). 

I also think that users should not expect they can mutate objects in Ruby's memory fetched by `read_attribute`, or accessed any other low-level ways such as `instance_variable_get(___)`. Instead, they should be required to use an official setter or `write_attribute`.

Reply:
I think I might agree but I would love to see what @p-mongo thinks